[Clang] Fix a crash when using ast-dump=json

elizabethandrews opened the pull request:
CXXDeductionGuideDecl inherits from FunctionDecl. For FunctionDecls, the JSONVisitor
includes a call to visit NamedDecl in order to provide mangled names in the dump. This did
not correctly exclude CXXDeductionGuideDecl, which resulted in an assert being hit.

tahonermann commented on clang/test/AST/ast-dump-template-decls-json.cpp:
```diff
@@ -217,21 +221,21 @@ void i();
 // CHECK-NEXT:    "id": "0x{{.*}}",
 // CHECK-NEXT:    "kind": "FunctionTemplateDecl",
 // CHECK-NEXT:    "loc": {
-// CHECK-NEXT:     "offset": {{[0-9]+}},
+// CHECK-NEXT:     "offset": 142,
```
It looks like a regular expression was previously used to avoid sensitivity to the `offset` values. I'm not familiar with this test, but I see the comment above regarding using `gen_ast_dump_json_test.py` to generate the `CHECK` lines. Did you use that script?

elizabethandrews replied:
Yes I did. All changes in this file where made by the script.

tahonermann replied:
Hmm, the regex changes were made in https://github.com/llvm/llvm-project/commit/6d8e5c968538eff2c024d46d5715de0828a69407 by Chris Bieneman and reviewed by Aaron in https://reviews.llvm.org/D119362. It looks like those regex changes need to be preserved. It seems that script could use an update. Perhaps follow up with Chris?

elizabethandrews replied:
It looks like all the script does is dump the output of `-ast-dump=json` without really checking the string. Theoretically I guess it would be possible to check for "offset" in this string and modifying what is generated, but I haven't worked with Python in years and so might not be the best person to do this update. I think this should also be done in a separate patch since it falls outside the scope of this patch. 

For now I can revert the offset changes the script made and just keep the new CHECK lines with offsets modified? Would you be ok with that? Tagging @llvm-beanz for any input as well.

AaronBallman replied:
I think we just forgot to update that script to produce regexes here; I think it's fine to revert back to the regex and fix the script in a follow-up (doesn't have to be @elizabethandrews doing the follow-up either).

llvm-beanz replied:
To add a bit of context on the change to using regex matches. These tests are fragile to your system’s native line-endings. I changed these to regex matches in an effort to get clang’s tests to pass under a git checkout with autocrlf enabled.

tahonermann replied:
Reverting the undesired changes suffices for me. My Python experience is years out of date too :)

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -377,6 +377,8 @@ Miscellaneous Clang Crashes Fixed
   `Issue 64065 <https://github.com/llvm/llvm-project/issues/64065>`_
 - Fixed a crash when check array access on zero-length element.
   `Issue 64564 <https://github.com/llvm/llvm-project/issues/64564>`_
+- Fixed a crash when -ast-dump=json was used for code using template
```
```suggestion
- Fixed a crash when ``-ast-dump=json`` was used for code using template
```

AaronBallman commented on clang/lib/AST/JSONNodeDumper.cpp:
```diff
@@ -819,6 +819,10 @@ void JSONNodeDumper::VisitNamedDecl(const NamedDecl *ND) {
       return;
 
+    // Do not mangle template deduction guides.
+    if (const auto DG = dyn_cast<CXXDeductionGuideDecl>(ND))
```
```suggestion
    if (isa<CXXDeductionGuideDecl>(ND))
```

elizabethandrews committed:
[Clang] Fix a crash when using ast-dump=json

CXXDeductionGuideDecl inherits from FunctionDecl. For FunctionDecls,
the JSONVisitor includes a call to visit NamedDecl in order to
provide mangled names in the dump. This did not correctly exclude
CXXDeductionGuideDecl, which resulted in an assert being hit.

elizabethandrews commented:
I apologize for the force push. I had a strange Git error I was trying to get rid of.  I also squashed my commits for the same reason. It doesn't look like it has helped though. I see the first test bot checking formatting fails with

```
Run python llvm/utils/git/code-format-helper.py \
Running: git-clang-format --diff aa30018e66105b4a8bd0947e35d1946d[24](https://github.com/llvm/llvm-project/actions/runs/6698182261/job/18199674028?pr=70224#step:8:25)2cd99c f4559a402d92fd6ab394b586b1c5f3f6d740c3a8 -- clang/lib/AST/JSONNodeDumper.cpp clang/test/AST/ast-dump-template-decls-json.cpp
error: 'aa30018e66105b4a8bd0947e35d1946d242cd99c' is not a commit
error: clang-format exited with code 2
error: some formatters failed: clang-format
Error: Process completed with exit code 1.
```

I have no idea why this is happening.

elizabethandrews commented:
I've applied all the review comments. I'm not sure what to do about the failing code format bot though. Should I open a new PR or  can I ignore that bot since I've run clang-format locally?

llvm-beanz commented:
@elizabethandrews ignore the clang-format failure. The action isn't fetching enough commits from git in some cases which leads to this failure.

elizabethandrews commented:
@tahonermann @AaronBallman does that sound ok to you ? Can I merge this (after a rebase)?

tahonermann commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -697,6 +697,8 @@ Miscellaneous Clang Crashes Fixed
   `Issue 64564 <https://github.com/llvm/llvm-project/issues/64564>`_
 - Fixed a crash when an ObjC ivar has an invalid type. See
   (`#68001 <https://github.com/llvm/llvm-project/pull/68001>`_)
+- Fixed a crash when ``-ast-dump=json`` was used for code using template
+  deduction guides.
```
```suggestion
- Fixed a crash when ``-ast-dump=json`` was used for code using class
  template deduction guides.
```

elizabethandrews replied:
Done!

tahonermann reviewed (approved):
I added a minor suggested change for the release note, but this otherwise looks good to me.
